dmoj/problem.py

```diff
@@ -45,6 +51,88 @@ def __init__(self, problem_id, time_limit, memory_limit):
             raise InvalidInitException(str(e))
 
         self.problem_data.archive = self._resolve_archive_files()
+        self._resolve_test_cases()
+
+    def _match_test_cases(self, filenames, input_case_pattern, output_case_pattern, case_points):
+        def try_match_int(match, group):
+            try:
+                val = match.group(group)
+            except IndexError:
+                return None
+
+            try:
+                return int(val)
+            except (ValueError, TypeError):
+                return val
+
+        def parse_position(pattern, filename):
+            match = pattern.match(filename)
+            if not match:
+                return None
+
+            # Allow batches and case numbers to be alphanumeric, in which case we will sort them lexicographically.
+            # Still attempt to process them as integers first, though, since most problems will use this format.
+            return try_match_int(match, 'batch'), try_match_int(match, 'case')
+
+        # Match all cases with the same (batch, position) mapping.
+        groups = defaultdict(list)
+        batch_ids = set()
+        for a in filenames:
```

Comment:
Can we not do O(n²)? This should not be rocket science for a grade 10 student. This code has to run for every submission...